lib/src/rules/directives_ordering.dart

```diff
@@ -229,6 +302,31 @@ class _Visitor extends SimpleAstVisitor {
         .forEach(reportDirective);
   }
 
+  void _checkSectionInOrder(
+      Set<AstNode> lintedNodes, Iterable<NamespaceDirective> nodes) {
+    void reportDirective(NamespaceDirective directive) {
+      if (lintedNodes.add(directive)) {
+        rule._reportLintWithDirectiveSectionOrderedAlphabeticallyMessage(
+            directive);
+      }
+    }
+
+    List<NamespaceDirective> addIfUpperBound(
+        List<NamespaceDirective> acc, NamespaceDirective curr) {
+      if (acc.isEmpty || acc.last.uriContent.compareTo(curr.uriContent) < 0) {
+        acc.add(curr);
+      }
+
+      return acc;
+    }
+
+    final inOrderNodes = nodes.fold(<NamespaceDirective>[], addIfUpperBound);
+
+    nodes.toSet()
+      ..removeAll(inOrderNodes)
+      ..forEach(reportDirective);
```

Comment:
Not everything needs to be done with a `fold` :-)

How about a simple loop:
```
NamespaceDirective previousDirective = null;
for (NamespaceDirective directive in nodes) {
  if (previousDirective == null || previousDirective.uriContent.compareTo(directive.uriContent) < 0) {
    previousDirective = directive;
  } else {
    reportDirective(directive);
  }
}
```
It isn't as terse, but it's a lot more performant.

Also, if directive d<sub>i</sub> should be before d<sub>j</sub> when i < j, and the directives being checked are in the order d<sub>1</sub>, d<sub>5</sub>, d<sub>2</sub>, d<sub>3</sub>, d<sub>4</sub>, then this algorithm will produce three lints. But if you rewrite the loop to:
```
NamespaceDirective previousDirective = null;
for (NamespaceDirective directive in nodes) {
  if (previousDirective != null && previousDirective.uriContent.compareTo(directive.uriContent) > 0) {
    reportDirective(directive);
  }
  previousDirective = directive;
}
```
then it will only produce one lint. I'd want to try it on some real data before concluding that this is better, but it's worth considering.